feat(cli): allow users to specify dynamic libraries directly

WillLillis opened the pull request:
### The Problem

The tree-sitter CLI does a lot of "magic" behind the scenes to find and load the correct dynamic library when various subcommands are used (i.e. `parse`, `query`, etc.). For some more advances use cases (i.e. testing two different commits for the same grammar), this auto-magical selection isn't desired.

### The Solution

This PR adds the ability to specify the path to a shared library directly via `--lib_path`. Internally, once the library has been loaded, the next step is to extract the language function symbol (i.e. `tree_sitter_c` from `c.so`). In my current implementation, we attempt to derive this symbol from the name of the shared library. If this doesn't work, users can pass `--language_name`.

Sample usage:

```
$ tree-sitter parse sample.c --lib-path /path/to/first/shared/lib/c.so
...
$ tree-sitter parse sample.c --lib-path /path/to/another/shared/lib/other_c.so --lang-name c
```

#### Other considerations

I'm opening this as a draft for now, until the following items can be settled agreed upon:

- [x] Ergonomics of the new arguments. Naming, documentation, etc.
- [x] Should a warning be displayed if `--lang_name` is specified when `--lib_path` isn't?
- [x] Further testing once I have some more time

Once we've settled on these changes, I'd like to:

- [x] Update the docs
- [x] *Potentially* extend this functionality to other subcommands (i.e. `fuzz`). Language loading works a bit differently for other subcommands, so we'll have to be careful not to introduce subtle regressions.

CC @sogaiu and @clason since you've both expressed interest in something like this.

Related: https://github.com/tree-sitter/tree-sitter/pull/4498

WillLillis commented on crates/loader/src/loader.rs:
```diff
@@ -1410,8 +1407,27 @@ impl Loader {
+        lib_info: Option<&(&Path, &str)>,
     ) -> Result<Language> {
-        if let Some(scope) = scope {
+        if let Some((lib_path, language_name)) = lib_info {
```
This logic is marginally adapted from the end of `Loader::load_language_at_path_with_name`. It may make sense to pull this logic into a helper for `Loader` so it can live in one place.

sogaiu commented:
I gave baebcea9 a brief try:

Here's a "log" of using `--lib-path`:
```
$ cd ~/src/tree-sitter-clojure

# make test file
$ echo "^:wake ^:sit ^:sleep #{}" > meta-on-set.clj

# build library
$ tree-sitter.load_shared build

# parse with --lib-path
$ tree-sitter.load_shared parse --cst --lib-path ~/src/tree-sitter-clojure/clojure.so meta-on-set.clj 
0:0  - 1:0    source 
0:0  - 0:24     set_lit 
0:0  - 0:6        meta: meta_lit 
0:0  - 0:1          "^"
0:1  - 0:6          value: kwd_lit 
0:1  - 0:2            ":"
0:2  - 0:6            name: kwd_name `wake`
0:7  - 0:12       meta: meta_lit 
0:7  - 0:8          "^"
0:8  - 0:12         value: kwd_lit 
0:8  - 0:9            ":"
0:9  - 0:12           name: kwd_name `sit`
0:13 - 0:20       meta: meta_lit 
0:13 - 0:14         "^"
0:14 - 0:20         value: kwd_lit 
0:14 - 0:15           ":"
0:15 - 0:20           name: kwd_name `sleep`
0:21 - 0:22       "#"
0:22 - 0:23       "{"
0:23 - 0:24       "}"
```

Here's a "log" in a different checkout, using `--lib-path` and then also `--language-name`:
```
$ cd /tmp

# clone grammar repository (output omitted)
$ git clone https://github.com/sogaiu/tree-sitter-clojure

$ cd tree-sitter-clojure

# checkout a different branch / commit from above
$ git checkout unstable-20250709

# make test file
$ echo "^:wake ^:sit ^:sleep #{}" > meta-on-set.clj

# build library
$ tree-sitter.load_shared build

# parse with --lib-path
$ tree-sitter.load_shared parse --cst --lib-path /tmp/tree-sitter-clojure/clojure.so meta-on-set.clj 
0:0  - 1:0    source 
0:0  - 0:6      meta: meta_lit 
0:0  - 0:1        "^"
0:1  - 0:6        value: kwd_lit 
0:1  - 0:2          ":"
0:2  - 0:6          name: kwd_name `wake`
0:7  - 0:12     meta: meta_lit 
0:7  - 0:8        "^"
0:8  - 0:12       value: kwd_lit 
0:8  - 0:9          ":"
0:9  - 0:12         name: kwd_name `sit`
0:13 - 0:20     meta: meta_lit 
0:13 - 0:14       "^"
0:14 - 0:20       value: kwd_lit 
0:14 - 0:15         ":"
0:15 - 0:20         name: kwd_name `sleep`
0:21 - 0:24     set_lit 
0:21 - 0:22       "#"
0:22 - 0:23       "{"
0:23 - 0:24       "}"

# prepare to test --language-name
$ cp clojure.so clj.so

# parse with --lib-path and --language-name
$ tree-sitter.load_shared parse --cst --lib-path /tmp/tree-sitter-clojure/clj.so --language-name clojure meta-on-set.clj 
0:0  - 1:0    source 
0:0  - 0:6      meta: meta_lit 
0:0  - 0:1        "^"
0:1  - 0:6        value: kwd_lit 
0:1  - 0:2          ":"
0:2  - 0:6          name: kwd_name `wake`
0:7  - 0:12     meta: meta_lit 
0:7  - 0:8        "^"
0:8  - 0:12       value: kwd_lit 
0:8  - 0:9          ":"
0:9  - 0:12         name: kwd_name `sit`
0:13 - 0:20     meta: meta_lit 
0:13 - 0:14       "^"
0:14 - 0:20       value: kwd_lit 
0:14 - 0:15         ":"
0:15 - 0:20         name: kwd_name `sleep`
0:21 - 0:24     set_lit 
0:21 - 0:22       "#"
0:22 - 0:23       "{"
0:23 - 0:24       "}"
```

The test file is being parsed differently as expected :)

Looks good so far :+1:

WillLillis commented:
> > Should a warning be displayed if --language_name is specified when --lib_path isn't?
> 
> Unless the value from `--language-name` can be used in some other situation where `--lib-path` is not specified, having a warning seems sensible.
> 
> On a somewhat related note, looking at `dump-languages` output, I don't see anything corresponding to values for `--language-name` (the closest being values for scope perhaps?). I wonder if it would be worth having the info there as well...

Added the warning, and good point on `dump-languages`!

sogaiu commented:
Warning seems to be working in f8704c6:

```
$ tree-sitter.load_shared parse --cst --language-name clojure meta-on-set.clj 
Warning: --language-name` specified without --lib-path. This argument will be ignored.
0:0  - 1:0    source 
0:0  - 0:24     set_lit 
0:0  - 0:6        meta: meta_lit 
0:0  - 0:1          "^"
0:1  - 0:6          value: kwd_lit 
0:1  - 0:2            ":"
0:2  - 0:6            name: kwd_name `wake`
0:7  - 0:12       meta: meta_lit 
0:7  - 0:8          "^"
0:8  - 0:12         value: kwd_lit 
0:8  - 0:9            ":"
0:9  - 0:12           name: kwd_name `sit`
0:13 - 0:20       meta: meta_lit 
0:13 - 0:14         "^"
0:14 - 0:20         value: kwd_lit 
0:14 - 0:15           ":"
0:15 - 0:20           name: kwd_name `sleep`
0:21 - 0:22       "#"
0:22 - 0:23       "{"
0:23 - 0:24       "}"
```

I see the new output from `dump-languages` as well (slightly edited):

```
$ tree-sitter.load_shared dump-languages
name: clojure
scope: source.clojure
parser: "$HOME/src/tree-sitter-clojure/."
highlights: None
file_types: ["bb", "clj", "cljc", "cljs", "edn"]
content_regex: None
injection_regex: None

name: clojure
scope: 
parser: "/tmp/tree-sitter-clojure"
highlights: None
file_types: []
content_regex: None
injection_regex: None
```

---

On the subject of ergnomics of the arguments...I wonder if folks would prefer `--lang-name` to `--language-name` [1].  We already seem to be abbreviating in the following cases:

* `config` for `configuration`
* `js` for `javascript` (in `--js-runtime`)
* `libdir` for `library directory` (?)
* `subdir` for `subdirectory`
* `stat` for `statistics`

This PR also currently uses `--lib-path` (which is fine by me :) ).

We're also using `grammar-path` so I wonder if there is some sense to `grammar-name`...though that's longer and we already say `dump-languages`...

---

[1] May be `lang` is not likely to be interpreted as anything other than `language`...especially in the context of tree-sitter?

WillLillis commented:
> I wonder if folks would prefer --lang-name to --language-name

Fine by me!

> We're also using grammar-path so I wonder if there is some sense to grammar-name...though that's longer and we already say dump-languages...

My main thought for using `grammar-path` there was because that argument points to a tree-sitter grammar directory. I'm not totally confident on the verbiage, but my sense was that the languages are something you "get out of" the grammars. Don't feel super strongly either way, but I'd lean towards keeping that one the same just for the sake of minimizing unnecessary breakage between versions.

sogaiu commented:
> My main thought for using grammar-path there was because that argument points to a tree-sitter grammar directory. 

Yes, I think using the word `grammar` (in `grammar-path`) is a fine choice.

Sorry, I didn't mean to imply that ought to change or anything.

There is also `name` inside of the `grammars` array in `tree-sitter.json` which I presume is meant to match what is in `grammar.js` as well as `grammar.json`.  Although I'm not highly enthusiastic about suggesting `grammar-name` over `lang-name`, I think `grammar-name` feels more consistent [1].

As someone who has worked on grammars for which there were multiple different implementations, the idea of there being multiple grammars for a single language is much more at the front of my mind so-to-speak.

Perhaps other folks have opinions though :)

---

[1] Though I won't be surprised to see mistypings of `grammar` being more common than `lang` :sweat_smile:

WillLillis commented:
> It occurs to me that since the `--lang-name` / `--grammar-name` option is not likely to be used as much:
> 
> > If this doesn't work, users can pass --language_name.
> 
> may be it's ok if the name is `--grammar-name`?

The only objection I'd have to this is that I found "language name" to be used internally while figuring this out:

https://github.com/tree-sitter/tree-sitter/blob/8e90799e2791a322f6df5c435280433e2c381773/crates/loader/src/loader.rs#L854-L859

Beyond that, I don't have a super strong preference either way. Going to give this some more thought as I work through the rest of this PR.

sogaiu commented:
> The only objection I'd have to this is that I found "language name" to be used internally for this kind of data

Since it's not likely to be used that often, may be in the interest of keeping things simpler, staying with `--language-name` or `--lang-name` is better then.

WillLillis commented:
>  Update the docs

I actually won't do this in *this* PR. The website is served off of the master branch, so it doesn't make sense to add this until we're ready to cut a release with these changes. I've added updates to some existing options that were missing entries, though. (Maybe this should be pulled into a separate PR that can be backported?)

> *Potentially* extend this functionality to other subcommands (i.e. fuzz). Language loading works a bit differently for other subcommands, so we'll have to be careful not to introduce subtle regressions.

I've added the `lib-path` and `--lang-name` arguments to the `test` and `fuzz` subcommands. I didn't add the arguments for the following subcommands:

- `highlight`: the current highlight functionality requires a `LanguageConfiguration` that is retrieved by the passed in `scope`. Hacking around this would lead to unintuitive/wrong behavior, and isn't work the added complexity imo.
- `tags` Similar story to `highlight`
- other subcommands: N/A
